### PR TITLE
LLaDA integration

### PR DESCRIPTION
**Pull request by brozonoyer**
## What

Integrate LLaDA (GSAI-ML/LLaDA-8B-Base/Instruct) into xlm-core at Dream parity: a backbone port with strict HF checkpoint loading, an `xlm-models/llada/` package with an MLM-protocol adapter and diffusion predictor, and Hydra configs including a MATH-500 eval experiment.

## Contribution type

- [x] Maintained model (`xlm-models/`)
- [ ] External model (separate repo)
- [ ] Task or dataset
- [x] Core framework (`src/xlm/`)
- [ ] Documentation
- [ ] Bug fix
- [ ] Breaking change (apply the `breaking-change` label and include migration notes below)
- [ ] Other (describe below)

## Labels

`model`, `enhancement`

## Changes

- **Backbone `src/xlm/backbones/llada/`**: `modeling_llada.py` / `configuration_llada.py` ported from the GSAI-ML/LLaDA-8B-Base trust_remote_code files, with state-dict keys kept identical to the checkpoint (`model.transformer.wte`, `model.transformer.blocks.*`, `model.transformer.ff_out`) so Hub weights load with `strict=True` and zero remapping — same approach as the Dream backbone.
  - Upstream `LLaDAConfig` renamed to `LLaDAConfigBase` (mirrors `DreamConfigBase`); `AutoConfig`/`AutoModel` registrations dropped to avoid clashing with the remote-code module in the same process.
  - Optional `position_ids` threaded through `LLaDAModelLM` → `LLaDAModel` → blocks → `RotaryEmbedding` (per-token RoPE positions, e.g. packed cumsum positions); default path is bitwise-unchanged.
  - transformers v5 compatibility: v5 no longer persists None-valued/`use_cache` config kwargs as attributes, so `LLaDAConfigBase.__init__` restores all `ModelConfig` fields explicitly (regression-tested). Note: `DreamConfig` likely has the same latent issue on v5 (e.g. `mask_token_id` silently dropped).
- **Package `xlm-models/llada/`**:
  - `llada_model.py`: `LLaDAXLMModel` MLM-protocol adapter (`forward(x_t, attention_mask, positions) -> logits`), analogue of `DreamXLMModel` — but no `LogitsShiftBy1` (LLaDA predicts masked positions in place) and no 4D mask expansion (backbone handles 2D masks natively).
  - `configuration_llada.py`: `LLaDAConfig(LLaDAConfigBase)` with an explicit `__init__` (required on transformers v5, which otherwise replaces subclass `__init__` with a generated dataclass one).
  - `predictor_llada.py`: `LLaDAPredictor(DreamPredictor)` — Dream's machinery already implements LLaDA's reference sampler (`confidence: top_prob` == `low_confidence` remasking, `block_size` == semi-AR `block_length`); mapping documented in the module docstring.
  - `datamodule_llada.py`: `print_batch_llada` debug helper.
- **Hydra configs `xlm-models/llada/configs/`**: `model/llada_8b.yaml` (one model file serves Base and Instruct; `hub.repo_id` selects the weights), `model_type/llada_base.yaml` + `llada_eval.yaml`, `experiment/math500_llada_eval.yaml` (tokenizer via `load_auto_tokenizer` with `mask_token: <|mdm_mask|>` → id 126336), plus datamodule/collator/debug/fsdp mirrors of Dream's.
- **Registration**: `"llada": "llada"` added to `xlm-models/xlm_models.json`; `llada` added to packages/package_dir/package_data in `xlm-models/setup.py`.
- **Tests `tests/models/llada/`**: 8 CPU tests (forward shape, MLM-protocol signature, position_ids parity, padding-mask effect, HF key layout, strict state-dict roundtrip, config-field roundtrip on v5) + an env-gated full-checkpoint parity test.

## Testing

```bash
# CPU tests (8 passed)
pytest tests/models/llada/test_llada_model.py -q

# Full 8B Hub checkpoint parity vs AutoModel(trust_remote_code=True):
# logits agree at atol/rtol 1e-3 in bf16 (downloads ~16GB; ~9 min on CPU)
XLM_RUN_HUB_TESTS=1 pytest tests/models/llada/test_llada_model.py -k hub

# Hydra composition of the eval experiment through the CLI (verified)
XLM_MODELS_PACKAGES="mlm:dream:llada" xlm job_name=check job_type=evaluate \
  experiment=math500_llada_eval --cfg job
  
  cc @dhruvdcoder @Durga-Prasad1

### CODE REVIEW

**Comment by brozonoyer** [comment]
## LLaDA integration

Ports the GSAI-ML LLaDA masked-diffusion LM into xlm-core:

- **Backbone** (`src/xlm/backbones/llada/`): HF modeling/configuration ported so `GSAI-ML/LLaDA-8B-{Base,Instruct}` checkpoints load with `strict=True` and no key remapping; adds 2D attention-mask and `position_ids` support. Logits match the `trust_remote_code` reference (parity test gated behind `XLM_RUN_HUB_TESTS=1`; CPU tests run by default).
- **Model package** (`xlm-models/llada/`): `LLaDAXLMModel` MLM-protocol adapter and `LLaDAPredictor`, a thin subclass of `DreamPredictor` — LLaDA's reference sampler maps onto existing machinery (`remasking='low_confidence'` ≡ `confidence: top_prob` + `confidence_temperature: 0`; `block_length` ≡ `block_size`). No `logits_hook`: LLaDA predicts masked positions in place (no Dream-style shift). `cfg_scale` is not implemented (unused in the base-model evals).
- **Configs/docs/tests**: `model/llada_8b`, `model_type/llada_{base,eval}`, MATH-500 eval experiment, docs page, registration in `xlm_models.json`/`setup.py`.

**Validation**: 4-shot MATH-500, LLaDA-8B-Base, reference base-model protocol (gen_length=256, steps=256, single block, greedy low-confidence remasking, temperature 0): **29.2% (124/424**; job hit its 12h wall limit, random subset, ±4.3pp @95%) vs ~33% reference; 0 `math_verify` parse failures.

**Shared-file changes**: (1) `DreamPredictor.predict` decodes only the generated span and trims at `generate_until` stop strings — previously the left-padded prompt leaked into logged predictions and post-hoc evaluators. (2) `MLMSeq2SeqPredCollator` passes `truncate` through to `prepare_prefix_ids` (was silently ignored; defaults unchanged).

**Known issue (pre-existing, worked around)**: the shared `default_hf` callbacks group references `xlm.datamodule.base.PrintBatchCallback` and `xlm.diffusion.base.UnconditionalSampleGenerator`, which no longer exist; the LLaDA experiment nulls them out. Needs a separate fix.
